modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java

```diff
@@ -99,17 +102,22 @@ public Map<String, Supplier<HttpServerTransport>> getHttpTransports(
         NamedXContentRegistry xContentRegistry,
         NetworkService networkService,
         HttpServerTransport.Dispatcher dispatcher,
+        BiConsumer<HttpPreRequest, ThreadContext> dispatcherContext,
         ClusterSettings clusterSettings,
         Tracer tracer
     ) {
+        HttpServerTransport.Dispatcher dispatcherWithContext = HttpServerTransport.Dispatcher.dispatchWithThreadContextWrapper(
+            dispatcher,
+            (httpPreRequest -> dispatcherContext.accept(httpPreRequest, threadPool.getThreadContext()))
+        );
```

Comment:
Previously, the `RestController` was in charge to build the thread context from the request headers.
Now, the `dispatcherContext` does that, and it is exposed by the `ActionModule`, to be used by the plugins that build `HttpServerTransport` implementations (which take the `RestController` as the argument).